### PR TITLE
Deployment pane

### LINKED ISSUE #959: Deployment pane
Currently, the output of compiler goes nowhere unless errored, the output of the flasher is shown right inside upload dialog, the debug log is put in the bottom pane which looks alien to the new sidebar system.

Let’s make things consistent and gather all logs in a single “Deployment” pane docked to bottom.

### Mockup

deploy

[deploy-psd.zip](https://github.com/xodio/xod/files/1568172/deploy-psd.zip)

Instead of the “context menu” button let’s use a simple accordion arrow to toggle pane visibility in a single click.

For filter targets lets use three scopes:

- Compiler output
- Flasher output
- Debugger protocol

Note, that errors and boundaries (“Compiling...”, “Flashing…”, “Debug session started…”, “Disconnected”) of all scopes are always added to the log.

### Icons

- [quick-upload.zip](https://github.com/xodio/xod/files/1568156/quick-upload.zip)
- [clear-log.zip](https://github.com/xodio/xod/files/1568169/clear-log.zip)
- [debug.zip](https://github.com/xodio/xod/files/1568170/debug.zip)
- [filter.zip](https://github.com/xodio/xod/files/1568171/filter.zip)

### Acceptance criteria

- [x] On the first IDE start the deployment pane is collapsed
- [x] Clicking zap button is equivalent to Deploy → Upload to Arduino
- [x] Clicking bug button is equivalent to Deploy → Upload to Arduino and checking “Debug after upload”
- [x] Eraser button clears the whole log data so that it contains 0 lines.
- [x] Filter submenu controls log rendering, not data and capturing
- [x] The progress bar shows up only during upload/flashing
- [x] If an error line is put in log, the pane is forced to expand
- [x] Error lines highlight in red
- [ ] When compilation/flash/debugger error happens a persistent snack bar message is shown with text like “Compilation failed \n See deployment log for error details”
- [x] When compilation/flash succeeds a transient snack bar message is shown with text like “Uploaded successfully”
- [x] If I scroll to the bottom of the log, the scroll becomes sticky
- [x] When I expand the pane, the scroll is sticky by default
- [x] “Transpiling…” is shown as a separate (the very first) log line during upload
- [x] The log lines are word-wrapped

### PR DESCRIPTION
**Pull request by evgenykochetkov**
Closes #959

### CODE REVIEW

**Comment by nkrkv** [comment]
:beetle: Snackbar messages are not affected by the deployment panel and show up z-above it effectively hiding its content

Expected behavior: snackbar messages are stacked y-above the panel in any of its states (collapsed/expanded)

**Comment by nkrkv** [comment]
> If I scroll to the bottom of the log, the scroll becomes sticky

Does not work for me

> When I expand the pane, the scroll is sticky by default

Does not work for me

**Comment by nkrkv** [comment]
The panel scrollbar is native and ugly. Could we use the styled one?

File: packages/xod-client/src/debugger/actionTypes.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+export const UPLOAD = 'UPLOAD';
+
 export const SHOW_DEBUGGER_PANEL = 'SHOW_DEBUGGER_PANEL';
 export const HIDE_DEBUGGER_PANEL = 'HIDE_DEBUGGER_PANEL';
 export const TOGGLE_DEBUGGER_PANEL = 'TOGGLE_DEBUGGER_PANEL';
```
Comment (nkrkv) [inline packages/xod-client/src/debugger/actionTypes.js]: Do we still need them?

**Comment by nkrkv** [comment]
From the issue:

> For filter targets lets use three scopes:
> 
> - Compiler output
> - Flasher output
> - Debugger protocol

I see compiler & flasher merged together in one item. Well, OK, maybe it makes sense. But I see no output from compiler at all. Only from the flasher. The compiler should output at least:

> Sketch uses 1486 bytes (4%) of program storage space. Maximum is 32256 bytes.
Global variables use 29 bytes (1%) of dynamic memory, leaving 2019 bytes for local variables. Maximum is 2048 bytes.

“Debugger” filter should not filter out `console-log` messages. The idea is: “We see the watched values anyway, should we clobber the log with them?” Filtering all debugger messages while being in the debug mode makes a little sense to me.

You understood it wrong. So, I guess, I explained it poorly and a xoder could misunderstand the menu item too. Maybe we should rename the filter from “Debugger” to “Watched Values”?

**Comment by nkrkv** [comment]
> “Transpiling…” is shown as a separate (the very first) logline during upload

Hmm… it is shown up, but only *after* transpiling is done. Maybe a little `setTimeout` to give a chance for the popup to hide and the logline to render would help?

File: packages/xod-client/src/debugger/containers/Debugger.jsx [hunk 2]
```diff
@@ -1,65 +1,165 @@
...
+          </span>
+
+          <div className="progress">
+            {uploadProgress}
+          </div>
+
+          <button
+            className="quick-upload-button"
+            onClick={onUploadClick}
+            title="Upload To Arduino"
```
Comment (nkrkv) [inline packages/xod-client/src/debugger/containers/Debugger.jsx]: Prepositions and articles are not capitalized: `Upload to Arduino`

File: packages/xod-client/src/debugger/containers/Debugger.jsx [hunk 3]
```diff
@@ -1,65 +1,165 @@
...
+          />
+          <button
+            className="debug-button"
+            onClick={onUploadAndDebugClick}
+            title="Upload And Debug"
```
Comment (nkrkv) [inline packages/xod-client/src/debugger/containers/Debugger.jsx]: `Upload and Debug`

**Comment by brusherru** [comment]
🐛  with long error messages and unexpected error 🤔 
image

Log:
```
{
  "childProcess": {},
  "code": 1,
  "errorCode": "UPLOAD_ERROR",
  "message": "Command failed: \"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/arduino-builder\" -hardware=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/hardware\" -hardware=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -libraries=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-libraries\" -tools=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/tools\" -tools=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -fqbn=\"arduino:avr:uno\" -build-path=\"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/build\" \"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp\"\n/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp: In function 'void xod::xod__core__console_log::evaluate(xod::Context)':\n/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1133:13: error: could not convert 'line' from 'xod::List<char>' to 'bool'\n     if (line) {\n             ^\nexit status 1\n `\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/arduino-builder\" -hardware=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/hardware\" -hardware=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -libraries=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-libraries\" -tools=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/tools\" -tools=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -fqbn=\"arduino:avr:uno\" -build-path=\"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/build\" \"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp\"` (exited with error code 1)",
  "name": "ChildProcessError",
  "stack": "Error: Command failed: \"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/arduino-builder\" -hardware=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/hardware\" -hardware=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -libraries=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-libraries\" -tools=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/tools\" -tools=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -fqbn=\"arduino:avr:uno\" -build-path=\"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/build\" \"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp\"\n/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp: In function 'void xod::xod__core__console_log::evaluate(xod::Context)':\n/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1133:13: error: could not convert 'line' from 'xod::List<char>' to 'bool'\n     if (line) {\n             ^\nexit status 1\n `\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/arduino-builder\" -hardware=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/hardware\" -hardware=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -libraries=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-libraries\" -tools=\"/Users/annies/Sites/xod/packages/xod-client-electron/arduino-builders/mac/tools\" -tools=\"/Users/annies/Library/Application Support/xod-client-electron/packages\" -fqbn=\"arduino:avr:uno\" -build-path=\"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/build\" \"/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp\"` (exited with error code 1)\n    at callback (/Users/annies/Sites/xod/node_modules/child-process-promise/lib/index.js:33:27)\n    at ChildProcess.exithandler (child_process.js:230:5)\n    at emitTwo (events.js:106:13)\n    at ChildProcess.emit (events.js:191:7)\n    at maybeClose (internal/child_process.js:885:16)\n    at Socket.<anonymous> (internal/child_process.js:334:11)\n    at emitOne (events.js:96:13)\n    at Socket.emit (events.js:188:7)\n    at Pipe._handle.close [as _onclose] (net.js:501:12)",
  "stderr": "/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp: In function 'void xod::xod__core__console_log::evaluate(xod::Context)':\n/Users/annies/Library/Application Support/xod-client-electron/upload-temp/xod-arduino-sketch.cpp:1133:13: error: could not convert 'line' from 'xod::List<char>' to 'bool'\n     if (line) {\n             ^\nexit status 1\n",
  "stdout": ""
}
```

**Comment by brusherru** [comment]
Also, about UX.

1. I think it would be handy if click on collapsed Pane will expand it (except clicks on icons).
2. And about design, in my opinion, a little odd that there are buttons which do not give visible effect in a collapsed state: clear log and filters.

@nkrkv what do you think about these two comments?

**Comment by brusherru** [comment]
🎨  One more thing about design.
I already implemented style for "check" icon when I do context menu for Sidebar Panels. I'm not sure that is perfect, but I think it's a little prettier.
image
To do this — just wrap "check" character into `<span className="state">✔</span>`.

Or at least add some space between this character and text :)

File: packages/xod-client/src/debugger/containers/Debugger.jsx [hunk 4]
```diff
@@ -1,65 +1,165 @@
...
+  [FILTER_TYPE.UPLOAD]: [UPLOAD_MSG_TYPE.FLASHER],
+  [FILTER_TYPE.DEBUG]: [UPLOAD_MSG_TYPE.LOG, UPLOAD_MSG_TYPE.XOD],
+};
+
+class Debugger extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      messageTypeFilter: { // TODO: better name?
```
Comment (brusherru) [inline packages/xod-client/src/debugger/containers/Debugger.jsx]: I think its ok and propose to remove TODO :)

File: packages/xod-client/src/debugger/containers/Debugger.jsx [hunk 5]
```diff
@@ -1,65 +1,165 @@
...
+            aria-valuenow={process}
+          />
+        </div>
+      ),
+      maybeUploadProgress
+    );
+
+    const { messageTypeFilter } = this.state;
+    const rejectedMessageTypes = R.compose(
+      R.chain(R.flip(R.prop)(FILTERED_MESSGE_TYPES)),
```
Comment (brusherru) [inline packages/xod-client/src/debugger/containers/Debugger.jsx]: `R.chain(R.prop(R.__, FILTERED_MESSAGE_TYPES)),`?

**Comment by nkrkv** [comment]
@brusherru @evgenykochetkov 

> I think it would be handy if click on collapsed Pane will expand it (except clicks on icons).

OK, reasonable. But only if there will be no gaps between buttons to avoid accidental misclicks.

> In my opinion, a little odd that there are buttons which do not give visible effect in a collapsed state: clear log and filters. Maybe we should hide it / disable or something else?

Good point. Let’s swap positions of uploading buttons and log control buttons. Then we could safely hide log control buttons in the collapsed state without any jumps.

**Comment by brusherru** [comment]
>> OK, reasonable. But only if there will be no gaps between buttons to avoid accidental misclicks.

To prevent it, a clickable zone could be to the left of buttons (wrapped with an additional clickable element), like this:
image
However, I am for the lack of a gap between the buttons, and they are not there :)

**Comment by evgenykochetkov** [comment]
All done except for native scroll, because it's handled by [`react-infinite`](https://github.com/seatgeek/react-infinite)

**Comment by nkrkv** [comment]
I’ve got a bunch of :bug: :bug: :bug: 

deployment-panel-glitches

- When starting a debug session I see a line of trash in the log after which it’s get cleared. I expect the last line to be “Debug session started” and the log kept intact after that (until I enable watched values output)
- Debugger breadcrumb panel is misaligned
- The sticky scroll is not sticky. When I scroll the log up a bit, it should *not* change anyhow until I stick it again (scroll to the bottom)
- Having `.Debugger .container` padding makes a little sense to me. Better remove it and give that 16px to the log: 180 → 196px

**Pull request merged**